UI: Add option to set fractional digits

CTNOriginals opened the pull request:
Adds a slider to the Numeric Display options page which effects the amount of fractional digits displayed in a regular number.
The slider allows a value between 0 and 5

Settings default:
<img width="810" height="298" alt="Settings menu" src="https://github.com/user-attachments/assets/fa753238-23d0-4fc8-966f-cf9e971335e8" />

0 fractional digits:
<img width="1116" height="657" alt="0 fractional digits:" src="https://github.com/user-attachments/assets/5f9e8e92-985b-4586-a3ee-671ae3d724fd" />

5 fractional digits:
<img width="1163" height="650" alt="5 fractional digits" src="https://github.com/user-attachments/assets/e41c527c-b535-461b-bf15-aaae2b7a28c0" />

### Possible edge case
Setting the option to 0 works as intended, the formatter removes the separator entirely and displays an integer instead of any amount of fractions behind it. 
Although, I did notice a minor edge case where the number sometimes displays a comma separated value like for example 1,000K when its really close to being a million, like 999995. 
I dont think this is caused by the changes applied in this fork.

### Relevant commits
- MISC: Add fractional digit settings field
- MISC: formatNumber fractionalDigits parameter default to Settings field
- UI: Add Fractional Digits option slider

### Credits
Thanks to @ficocelliguy for explaining the ropes of how to set up a PR.
Thanks to @shyguy1412 for being there at the right time with right answer to something very specific.

CTNOriginals committed:
MISC: Add fractional digit settings field

Prepare the settings object to later add the functionality

CTNOriginals committed:
MISC: formatNumber fractionalDigits parameter default to Settings field

Changed the default parameter for fractionalDigits to Settings.fractionalDigits.
Once this field is altered in settings, the function will simply apply it if nothing else was provided. This will include most numbers in the ui that are not of any special type (like percentages or storage sizes).

CTNOriginals committed:
UI: Add Fractional Digits option slider

This adds a slider to the Numeric Display options page that is tied to the fractionalDigits field in Settings.

Setting this to 0 works as intended, the formatter removes the seperator entirely and displays an integer instead of any amount of fractions behind it. Although, I did notice a minor edge case where the number sometimes displays a comma separated value like for example 1,000K when its really close to being a million, like 999995. I dont think this is caused by the changes applied in this fork..

CTNOriginals committed:
MISC: Fixed wording in option tooltip

As recommended by @ficocelliguy on discord

Co-Authored-By: Michael Ficocelli <1338468+ficocelliguy@users.noreply.github.com>

CTNOriginals committed:
MISC: npm format

forgot to run npm format.

CTNOriginals committed:
FIX: Lowerd the max fractional digits to prevent overflow

I noticed that the ui may seem a little off as the numbers overflow at some places and 5 didnt seem to make anything really overflow to disturbingly anymore anywhere that i could find.

d0sboots commented on src/GameOptions/ui/NumericDisplayOptions.tsx:
```diff
@@ -72,6 +77,15 @@ export const NumericDisplayPage = (): React.ReactElement => {
           <>If this is set all references to memory will use GiB instead of GB, in accordance with IEC 60027-2.</>
         }
       />
+      <OptionsSlider
```
I would move this up to between "Hide thousands separator" and "Hide trailing fractional zeroes for decimals". I think it fits best there? (Basically, grouping similar settings together, and introduce the idea of "fractional digits" before the more specific setting of "hide zeroes")

CTNOriginals replied:
Ah yeah fair.
I either wanted to do that, or just place myself near the bottom.
But I agree it would be nicer to group things but it does look at little inconsistent as all other options are toggles.

d0sboots commented on src/GameOptions/ui/NumericDisplayOptions.tsx:
```diff
@@ -72,6 +77,15 @@ export const NumericDisplayPage = (): React.ReactElement => {
+        step={1}
+        min={0}
+        max={5}
+        tooltip={<>The number of digits of precision to display on small numbers.</>}
```
```suggestion
        tooltip={<>The default number of decimal places to display on small numbers.</>}
```
"Decimal places"/"fractional digits" are not the same as "digits of precision". 12345.6789 to 3 fractional digits is 12345.679. To 3 digits of precision, it is 12300.

IMO also useful to mention "default", since lots of methods override this and thus won't be affected by this setting.

CTNOriginals replied:
Oh man you would not believe how long i was stuck on the wording of this, usually im fine with it but yesterday was not an english language day i guess. thanks for the improvement.

d0sboots commented on src/GameOptions/ui/NumericDisplayOptions.tsx:
```diff
@@ -4,10 +4,15 @@ import { Settings } from "../../Settings/Settings";
+  function handleFractionalDigitChange(_event: Event | React.SyntheticEvent, newValue: number | number[]): void {
+    Settings.fractionalDigits = newValue as number;
+    FormatsNeedToChange.emit();
+  }
```
```suggestion
  }
  
```
Style: Newline between functions.

d0sboots commented on src/Settings/Settings.ts:
```diff
@@ -192,6 +192,8 @@ export const Settings = {
   useEngineeringNotation: false,
   /** Whether to disable suffixes and always use exponential form (scientific or engineering). */
   disableSuffixes: false,
+  /** The amount of digits displayed after the period '.' in a number. */
```
```suggestion
  /** The default amount of digits displayed after the period '.' in a number. */
```

ficocelliguy commented:
That's a VERY long slider XD
You can probably cap it at like 300px width or something?

d0sboots commented:
I thought the same, but it's consistent with settings in other tabs :)

CTNOriginals committed:
Update src/GameOptions/ui/NumericDisplayOptions.tsx

Co-authored-by: David Walker <d0sboots@gmail.com>

CTNOriginals committed:
Update src/Settings/Settings.ts

Co-authored-by: David Walker <d0sboots@gmail.com>

CTNOriginals committed:
Update src/GameOptions/ui/NumericDisplayOptions.tsx

Co-authored-by: David Walker <d0sboots@gmail.com>

catloversg commented on src/Settings/Settings.ts:
```diff
@@ -192,6 +192,8 @@ export const Settings = {
   useEngineeringNotation: false,
   /** Whether to disable suffixes and always use exponential form (scientific or engineering). */
   disableSuffixes: false,
+  /** The default amount of digits displayed after the period '.' in a number. */
```
Optional suggestion:
```suggestion
  /** The default amount of digits displayed after the decimal separator. */
```
"Decimal point" and "decimal comma" are more common terms, but their usage depends on countries.

catloversg reviewed (commented):
You also need to rerun `npm run format` after applying other suggestions.

CTNOriginals committed:
MISC: Move setting & update tooltip

Moved the setting to the suggestion location and updated the tooltip to include the default value.

CTNOriginals committed:
MISC: npm run format & lint

CTNOriginals committed:
MISC: Update settings comment

Co-authored-by: catloversg <152669316+catloversg@users.noreply.github.com>
